### PR TITLE
Remove Lexer#reset method

### PR DESCRIPTION
**Pull request by sferik**
…as discussed in https://github.com/YorickPeterse/oga/issues/153.

### CODE REVIEW

**Comment by sferik** [comment]
@YorickPeterse It looks like this pull request has broken a few things. Any idea what I’m doing wrong here.

**Comment by sferik** [comment]
@YorickPeterse Never mind. I think I figured out what I was doing wrong. Please review and let me know what you think.

**Comment by sferik** [comment]
@YorickPeterse Looks like tests pass on everything except Rubinius on OS X. I guess that’s a spurious failure unrelated to this change.

File: Gemfile [hunk 1]
```diff
@@ -2,6 +2,10 @@ source 'https://rubygems.org'
 
 gemspec
 
+platforms :ruby_19 do
+  gem 'json', '~> 1.8'
+end
+
```
Comment (YorickPeterse) [inline Gemfile]: Why is this change needed? Oga doesn't even use the JSON Gem (at least directly) in any way.

File: lib/oga/xml/parser.rll [hunk 2]
```diff
@@ -245,7 +245,7 @@ string_body
   def reset
     @line = 1
 
-    @lexer.reset
+    @lexer.reset_native
   end
```
Comment (YorickPeterse) [inline lib/oga/xml/parser.rll]: This `reset` method can also just be removed as parser instances aren't re-used.

File: lib/oga/xml/lexer.rb [hunk 3]
```diff
@@ -161,7 +147,7 @@ def lex
           tokens << [type, value, line]
         end
 
-        reset
+        reset_native
```
Comment (YorickPeterse) [inline lib/oga/xml/lexer.rb]: This can be removed as well, `reset_native` should only be called in `initialize`.

**Comment by YorickPeterse** [comment]
@sferik The Rbx build indeed seems like some random timeout related problem. However, all Windows builds are failing as well.

File: Gemfile (hunk 1 above)
Comment (sferik) [inline Gemfile]: Version of the `json` gem greater than or equal to 2.0.0 are not compatible with Ruby 1.9, causing a build error on the Travis CI. This allows Travis to get past the `bundle` phase and run the tests successfully on Ruby 1.9.

File: lib/oga/xml/parser.rll (hunk 2 above)
Comment (sferik) [inline lib/oga/xml/parser.rll]: I originally removed this method but that caused test failures locally, so I added it back.

**Comment by sferik** [comment]
> However, all Windows builds are failing as well.

@YorickPeterse It looks like the Windows builds are only failing on Ruby 1.9 for the same reason that the Linux builds were failing on Ruby 1.9: incompatibility with version 2.0.0 or greater of the `json` gem.

**Comment by sferik** [comment]
@YorickPeterse Okay. All the tests seem to be passing now, even on Windows.

File: Gemfile (hunk 1 above)
Comment (YorickPeterse) [inline Gemfile]: @sferik Then I'd rather just drop 1.9 from Travis/AppVeyor, I think it's officially end of life anyway.

File: lib/oga/xml/parser.rll (hunk 2 above)
Comment (YorickPeterse) [inline lib/oga/xml/parser.rll]: @sferik If the code of the `reset` method is moved into the `initialize` method I think it should work. Which tests were failing?

File: Gemfile (hunk 1 above)
Comment (YorickPeterse) [inline Gemfile]: @sferik Oh hm, I see it affects Windows as well. Lets just keep it in the Gemfile for now.

File: lib/oga/xml/parser.rll (hunk 2 above)
Comment (sferik) [inline lib/oga/xml/parser.rll]: @YorickPeterse Removing this causes 6 tests to fail. I just pushed [this change](https://github.com/YorickPeterse/oga/pull/154/commits/08df2d2340fc7bdbe36b14cbf52ae96bd7451863) so you can see the tests failures on [Travis CI](https://travis-ci.org/YorickPeterse/oga/builds/143612168). Here is a snippet of the output:

```
Failures:

  1) Oga::XML::PullParser tracking nodes resets the current node after parsing
     Failure/Error: nesting << @node.name
     
     NoMethodError:
       undefined method `<<' for nil:NilClass
     # ./lib/oga/xml/pull_parser.rb:142:in `on_element'
     # ./lib/oga/xml/parser.rb:491:in `_rule_27'
     # ./lib/oga/xml/parser.rb:232:in `block in each_token'
     # ./lib/oga/xml/lexer.rb:219:in `add_token'
     # ./lib/oga/xml/lexer.rb:421:in `on_element_end'
     # ./lib/oga/xml/lexer.rb:172:in `advance_native'
     # ./lib/oga/xml/lexer.rb:172:in `block in advance'
     # ./lib/oga/xml/lexer.rb:126:in `read_data'
     # ./lib/oga/xml/lexer.rb:171:in `advance'
     # ./lib/oga/xml/parser.rb:229:in `each_token'
     # ./lib/oga/xml/pull_parser.rb:72:in `parse'
     # ./lib/oga/xml/pull_parser.rb:72:in `parse'
     # ./spec/oga/xml/pull_parser/general_spec.rb:16:in `block (3 levels) in <top (required)>'

  2) Oga::XML::PullParser tracking nodes tracks the current node
     Failure/Error: raise LL::ParserError, message
     
     LL::ParserError:
       Unexpected end of input, expected element closing tag instead on line 1
     # ./lib/oga/xml/parser.rb:257:in `parser_error'
     # ./lib/oga/xml/parser.rb:235:in `each_token'
     # ./lib/oga/xml/pull_parser.rb:72:in `parse'
     # ./lib/oga/xml/pull_parser.rb:72:in `parse'
     # ./spec/oga/xml/pull_parser/general_spec.rb:10:in `block (3 levels) in <top (required)>'

  3) Oga::XML::PullParser elements parses the text of an element
     Failure/Error: nesting << @node.name
     
     NoMethodError:
       undefined method `<<' for nil:NilClass
     # ./lib/oga/xml/pull_parser.rb:142:in `on_element'
     # ./lib/oga/xml/parser.rb:491:in `_rule_27'
     # ./lib/oga/xml/parser.rb:232:in `block in each_token'
     # ./lib/oga/xml/lexer.rb:219:in `add_token'
     # ./lib/oga/xml/lexer.rb:431:in `on_text'
     # ./lib/oga/xml/lexer.rb:172:in `advance_native'
     # ./lib/oga/xml/lexer.rb:172:in `block in advance'
     # ./lib/oga/xml/lexer.rb:126:in `read_data'
     # ./lib/oga/xml/lexer.rb:171:in `advance'
     # ./lib/oga/xml/parser.rb:229:in `each_token'
     # ./lib/oga/xml/pull_parser.rb:72:in `parse'
     # ./lib/oga/xml/pull_parser.rb:72:in `parse'
     # ./spec/oga/xml/pull_parser/element_spec.rb:22:in `block (3 levels) in <top (required)>'

  4) Oga::XML::PullParser elements parses an element
     Failure/Error: raise LL::ParserError, message
     
     LL::ParserError:
       Unexpected end of input, expected element closing tag instead on line 1
     # ./lib/oga/xml/parser.rb:257:in `parser_error'
     # ./lib/oga/xml/parser.rb:235:in `each_token'
     # ./lib/oga/xml/pull_parser.rb:72:in `parse'
     # ./lib/oga/xml/pull_parser.rb:72:in `parse'
     # ./spec/oga/xml/pull_parser/element_spec.rb:12:in `block (3 levels) in <top (required)>'

  5) Oga::XML::PullParser tracking element nesting sets the nesting for the outer element
     Failure/Error: nesting << @node.name
     
     NoMethodError:
       undefined method `<<' for nil:NilClass
     # ./lib/oga/xml/pull_parser.rb:142:in `on_element'
     # ./lib/oga/xml/parser.rb:491:in `_rule_27'
     # ./lib/oga/xml/parser.rb:232:in `block in each_token'
     # ./lib/oga/xml/lexer.rb:219:in `add_token'
     # ./lib/oga/xml/lexer.rb:380:in `add_element'
     # ./lib/oga/xml/lexer.rb:351:in `on_element_name'
     # ./lib/oga/xml/lexer.rb:172:in `advance_native'
     # ./lib/oga/xml/lexer.rb:172:in `block in advance'
     # ./lib/oga/xml/lexer.rb:126:in `read_data'
     # ./lib/oga/xml/lexer.rb:171:in `advance'
     # ./lib/oga/xml/parser.rb:229:in `each_token'
     # ./lib/oga/xml/pull_parser.rb:72:in `parse'
     # ./lib/oga/xml/pull_parser.rb:72:in `parse'
     # ./spec/oga/xml/pull_parser/element_nesting_spec.rb:10:in `block (3 levels) in <top (required)>'

  6) Oga::XML::PullParser tracking element nesting pops element names after leaving an element
     Failure/Error: nesting << @node.name
     
     NoMethodError:
       undefined method `<<' for nil:NilClass
     # ./lib/oga/xml/pull_parser.rb:142:in `on_element'
     # ./lib/oga/xml/parser.rb:491:in `_rule_27'
     # ./lib/oga/xml/parser.rb:232:in `block in each_token'
     # ./lib/oga/xml/lexer.rb:219:in `add_token'
     # ./lib/oga/xml/lexer.rb:380:in `add_element'
     # ./lib/oga/xml/lexer.rb:351:in `on_element_name'
     # ./lib/oga/xml/lexer.rb:172:in `advance_native'
     # ./lib/oga/xml/lexer.rb:172:in `block in advance'
     # ./lib/oga/xml/lexer.rb:126:in `read_data'
     # ./lib/oga/xml/lexer.rb:171:in `advance'
     # ./lib/oga/xml/parser.rb:229:in `each_token'
     # ./lib/oga/xml/pull_parser.rb:72:in `parse'
     # ./lib/oga/xml/pull_parser.rb:72:in `parse'
     # ./spec/oga/xml/pull_parser/element_nesting_spec.rb:20:in `block (3 levels) in <top (required)>'

Finished in 0.75791 seconds (files took 0.53493 seconds to load)
1554 examples, 6 failures

Failed examples:

rspec ./spec/oga/xml/pull_parser/general_spec.rb:15 # Oga::XML::PullParser tracking nodes resets the current node after parsing
rspec ./spec/oga/xml/pull_parser/general_spec.rb:9 # Oga::XML::PullParser tracking nodes tracks the current node
rspec ./spec/oga/xml/pull_parser/element_spec.rb:19 # Oga::XML::PullParser elements parses the text of an element
rspec ./spec/oga/xml/pull_parser/element_spec.rb:9 # Oga::XML::PullParser elements parses an element
rspec ./spec/oga/xml/pull_parser/element_nesting_spec.rb:9 # Oga::XML::PullParser tracking element nesting sets the nesting for the outer element
rspec ./spec/oga/xml/pull_parser/element_nesting_spec.rb:17 # Oga::XML::PullParser tracking element nesting pops element names after leaving an element
```

Comment (YorickPeterse) [inline lib/oga/xml/parser.rll]: @sferik Ah, I see the problem now. The class `Oga::XML::PullParser` also has a `reset` method that calls `super`. The code of said method can also be moved to `PullParser#initialize`, then the call to `reset` can be removed from `PullParser#parse` (the `parse` method itself here should stay).

Comment (sferik) [inline lib/oga/xml/parser.rll]: That seems to have fixed it. I’ve pushed these changes and the tests all seem to pass now.

**Comment by sferik** [comment]
@YorickPeterse Rubinius is failing nondeterministically. I’m going add it to the allowed failures list.

**Comment by sferik** [comment]
@YorickPeterse I was able to shave ~4 minutes off the Travis CI build time just by changing the order:

screen_shot_2016-07-10_at_12_36_29_pm

**Comment by YorickPeterse** [comment]
@sferik Please submit any Travis CI changes as a separate PR. If rbx keeps failing for reasons unknown I'll just merge this MR.

**Comment by sferik** [comment]
@YorickPeterse I’ve moved the changes to `.travis.yml` into [two](https://github.com/YorickPeterse/oga/pull/155) [separate](https://github.com/YorickPeterse/oga/pull/156) pull requests. The build for this pull request is now failing for Rubinius on macOS but passing everywhere else:

screen shot 2016-07-12 at 11 58 45 am

screen shot 2016-07-12 at 12 23 14 pm

**Comment by YorickPeterse** [comment]
I rebased this into master (latest commit is 3a89dcffab11977ad1edfbb1b0b00eb7d4a8712e), thanks!

**Comment by sferik** [comment]
Thank you for helping me put together the patch. I’ve run the `multi_xml` test suite against `oga@master` and everything is green. 💚

**Comment by YorickPeterse** [comment]
@sferik These changes are available in Oga 2.3 which I just pushed to RubyGems.

**Comment by sferik** [comment]
@YorickPeterse https://travis-ci.org/sferik/multi_xml/builds/144577237
